### DISCUSSION TITLE
v3.0.1-rc1

### DISCUSSION THREAD

**Discussion by mleone87** [Telmate/terraform-provider-proxmox#910] [Announcements]
Ladies and gentlemen, the new release is almost ready!!

This RC1.

This version introduces breaking changes regarding the disks layout! Refer to documentation for further explanations 

Please submit your feedbacks!

thanks to @Tinyblargon for the work on API side and on the provider side!

## What's Changed
* Manage issues stale by @mleone87 in https://github.com/Telmate/terraform-provider-proxmox/pull/724
* docs: Remove deprecated preprovision section by @kencx in https://github.com/Telmate/terraform-provider-proxmox/pull/733
* fixes #712 : destroy container fails "500 CT not running" by @btorresgil in https://github.com/Telmate/terraform-provider-proxmox/pull/728
* Add new vm_state attribute to manage running/stopped VMs by @jonglezb in https://github.com/Telmate/terraform-provider-proxmox/pull/725
* Smbios support for Qemu guest VM by @lsjostro in https://github.com/Telmate/terraform-provider-proxmox/pull/736
* fix: Ignore disk resize if media is cdrom by @lsjostro in https://github.com/Telmate/terraform-provider-proxmox/pull/737
* feat: update to go 1.19 by @mleone87 in https://github.com/Telmate/terraform-provider-proxmox/pull/740
* change PXE boot order and clean by @GannonTdW in https://github.com/Telmate/terraform-provider-proxmox/pull/729
* Bump github.com/rs/zerolog from 1.29.0 to 1.29.1 by @dependabot in https://github.com/Telmate/terraform-provider-proxmox/pull/746
* various stability & performance fixes by @mleone87 in https://github.com/Telmate/terraform-provider-proxmox/pull/749
* Update documentation with missing parameters for vm_qemu disk configuration by @AnotherIdea in https://github.com/Telmate/terraform-provider-proxmox/pull/756
* feat: cloud init disk resource by @lsjostro in https://github.com/Telmate/terraform-provider-proxmox/pull/748
* lxc: implement deletion of network interfaces by @GamerBene19 in https://github.com/Telmate/terraform-provider-proxmox/pull/754
* fix crash when disk.file is nil by @angelnu in https://github.com/Telmate/terraform-provider-proxmox/pull/765
* Add disk parameters serial and wwn by @sergelogvinov in https://github.com/Telmate/terraform-provider-proxmox/pull/767
* docs: add smbios to qemu docs + example by @nepeat in https://github.com/Telmate/terraform-provider-proxmox/pull/776
* Bump github.com/kdomanski/iso9660 from 0.3.3 to 0.3.5 by @dependabot in https://github.com/Telmate/terraform-provider-proxmox/pull/780
* Bump github.com/google/uuid from 1.3.0 to 1.3.1 by @dependabot in https://github.com/Telmate/terraform-provider-proxmox/pull/834
* Bump github.com/kdomanski/iso9660 from 0.3.5 to 0.4.0 by @dependabot in https://github.com/Telmate/terraform-provider-proxmox/pull/835
* Bump github.com/rs/zerolog from 1.29.1 to 1.31.0 by @dependabot in https://github.com/Telmate/terraform-provider-proxmox/pull/845
* Bump github.com/hashicorp/terraform-plugin-sdk/v2 from 2.26.1 to 2.29.0 by @dependabot in https://github.com/Telmate/terraform-provider-proxmox/pull/842
* feat: Add 'import-from' parameter to qemu by @jamesharr in https://github.com/Telmate/terraform-provider-proxmox/pull/606
* Bump golang.org/x/net from 0.9.0 to 0.17.0 by @dependabot in https://github.com/Telmate/terraform-provider-proxmox/pull/873
* Bump google.golang.org/grpc from 1.57.0 to 1.57.1 by @dependabot in https://github.com/Telmate/terraform-provider-proxmox/pull/875
* Add initial efidisk support by @riverar in https://github.com/Telmate/terraform-provider-proxmox/pull/732
* don't reboot on update if automatic_reboot is disabled by @quantumsheep in https://github.com/Telmate/terraform-provider-proxmox/pull/759
* Added multiple target nodes instead of a single node. by @PascalKu in https://github.com/Telmate/terraform-provider-proxmox/pull/766
* Fix Go module name (#790) by @pandatix in https://github.com/Telmate/terraform-provider-proxmox/pull/791
* Adding the role SND.use required for lxc in Proxmox 8.0.3 by @hnrazevedo in https://github.com/Telmate/terraform-provider-proxmox/pull/809
* Allow the creation of bind mounts for lxc containers by @zinefer in https://github.com/Telmate/terraform-provider-proxmox/pull/806
* added proxmox8 specific permission by @thigg in https://github.com/Telmate/terraform-provider-proxmox/pull/820
* Add ISO Storage resource by @h0lyalg0rithm in https://github.com/Telmate/terraform-provider-proxmox/pull/822
* Make cores not default to 1 to allow for "unlimited" cores on containers by @devzeebo in https://github.com/Telmate/terraform-provider-proxmox/pull/829
* Fix lxc container size: Accept K by @Mice7R in https://github.com/Telmate/terraform-provider-proxmox/pull/836
* add validation to vm_qemu name field by @hedenface in https://github.com/Telmate/terraform-provider-proxmox/pull/858
* Bugfix: reboot VM gracefully by @lucian-tx in https://github.com/Telmate/terraform-provider-proxmox/pull/861
* Feat: Gracefully shutdown VM on deletion. by @t3hpr1m3 in https://github.com/Telmate/terraform-provider-proxmox/pull/872
* allow using dynamic list for ipconfig instead of hardcoding the index in the key by @spettinichi in https://github.com/Telmate/terraform-provider-proxmox/pull/866
* Bump golang.org/x/crypto from 0.16.0 to 0.17.0 by @dependabot in https://github.com/Telmate/terraform-provider-proxmox/pull/881
* Bump github.com/google/uuid from 1.4.0 to 1.5.0 by @dependabot in https://github.com/Telmate/terraform-provider-proxmox/pull/880
* Bump github.com/hashicorp/terraform-plugin-sdk/v2 from 2.30.0 to 2.31.0 by @dependabot in https://github.com/Telmate/terraform-provider-proxmox/pull/879
* Bugfix/require recent module by @MaartendeKruijf in https://github.com/Telmate/terraform-provider-proxmox/pull/878
* resource_vm_qemu: allow . (dot) in name of virtual machine by @vilhelmprytz in https://github.com/Telmate/terraform-provider-proxmox/pull/886
* docs: New developer getting started guide by @hestiahacker in https://github.com/Telmate/terraform-provider-proxmox/pull/888
* Bump github.com/cloudflare/circl from 1.3.3 to 1.3.7 by @dependabot in https://github.com/Telmate/terraform-provider-proxmox/pull/894
* docs: added instructions on how to use a local library by @hestiahacker in https://github.com/Telmate/terraform-provider-proxmox/pull/890
* feat: overhaul qemu disks implementation by @Tinyblargon in https://github.com/Telmate/terraform-provider-proxmox/pull/892
* chore: update modules by @mleone87 in https://github.com/Telmate/terraform-provider-proxmox/pull/899
* Revert d02bb5d not properly tested by @mleone87 in https://github.com/Telmate/terraform-provider-proxmox/pull/905
* chore: update dependencies by @mleone87 in https://github.com/Telmate/terraform-provider-proxmox/pull/906

## New Contributors
* @kencx made their first contribution in https://github.com/Telmate/terraform-provider-proxmox/pull/733
* @btorresgil made their first contribution in https://github.com/Telmate/terraform-provider-proxmox/pull/728
* @jonglezb made their first contribution in https://github.com/Telmate/terraform-provider-proxmox/pull/725
* @lsjostro made their first contribution in https://github.com/Telmate/terraform-provider-proxmox/pull/736
* @AnotherIdea made their first contribution in https://github.com/Telmate/terraform-provider-proxmox/pull/756
* @GamerBene19 made their first contribution in https://github.com/Telmate/terraform-provider-proxmox/pull/754
* @angelnu made their first contribution in https://github.com/Telmate/terraform-provider-proxmox/pull/765
* @sergelogvinov made their first contribution in https://github.com/Telmate/terraform-provider-proxmox/pull/767
* @nepeat made their first contribution in https://github.com/Telmate/terraform-provider-proxmox/pull/776
* @riverar made their first contribution in https://github.com/Telmate/terraform-provider-proxmox/pull/732
* @quantumsheep made their first contribution in https://github.com/Telmate/terraform-provider-proxmox/pull/759
* @PascalKu made their first contribution in https://github.com/Telmate/terraform-provider-proxmox/pull/766
* @pandatix made their first contribution in https://github.com/Telmate/terraform-provider-proxmox/pull/791
* @hnrazevedo made their first contribution in https://github.com/Telmate/terraform-provider-proxmox/pull/809
* @zinefer made their first contribution in https://github.com/Telmate/terraform-provider-proxmox/pull/806
* @thigg made their first contribution in https://github.com/Telmate/terraform-provider-proxmox/pull/820
* @h0lyalg0rithm made their first contribution in https://github.com/Telmate/terraform-provider-proxmox/pull/822
* @devzeebo made their first contribution in https://github.com/Telmate/terraform-provider-proxmox/pull/829
* @Mice7R made their first contribution in https://github.com/Telmate/terraform-provider-proxmox/pull/836
* @hedenface made their first contribution in https://github.com/Telmate/terraform-provider-proxmox/pull/858
* @lucian-tx made their first contribution in https://github.com/Telmate/terraform-provider-proxmox/pull/861
* @t3hpr1m3 made their first contribution in https://github.com/Telmate/terraform-provider-proxmox/pull/872
* @spettinichi made their first contribution in https://github.com/Telmate/terraform-provider-proxmox/pull/866
* @MaartendeKruijf made their first contribution in https://github.com/Telmate/terraform-provider-proxmox/pull/878
* @vilhelmprytz made their first contribution in https://github.com/Telmate/terraform-provider-proxmox/pull/886
* @hestiahacker made their first contribution in https://github.com/Telmate/terraform-provider-proxmox/pull/888

**Full Changelog**: https://github.com/Telmate/terraform-provider-proxmox/compare/v2.9.14...v3.0.1-rc1

<hr /><em>This discussion was created from the release <a href='https://github.com/Telmate/terraform-provider-proxmox/releases/tag/v3.0.1-rc1'>v3.0.1-rc1</a>.</em>

**Comment by hestiahacker** [comment]
I've done about 5 hours of testing and do think this release candidate should be promoted to an official release.

I have confirmed that there are some major issue with permissions in this release (namely #784 and #687), but they are also in the latest stable release, which is why I don't think it should hold up an official version 3.0.1.

**Comment by Jasmont** [comment]
In 2.9.14 there was an option for specifying volume for disks in proxmox_vm_qemu resource. Couldn't find it in this release. Are there any plans to add it?

**Reply by sajjadG** [reply]
That's the breaking change mentioned in the first post. Check the doc https://github.com/Telmate/terraform-provider-proxmox/blob/master/docs/resources/vm_qemu.md#disks-block

**Comment by electropolis** [comment]
I think unless there will be no resolution with disappearing of cloudinit drive that provides NOT a workaround but a stable solution that works with `cicustom` also, which I mention here https://github.com/Telmate/terraform-provider-proxmox/issues/956 and attached my setup of `cicustom` to it by adjusting it to new requirements of this RC version (removed deprecated components), there should not be a `stable` release yet as it's for now still dropping cloudinit drive no matter what kind of workaround is added.

**Reply by hestiahacker** [reply]
It looks like this was fixed in the default branch, but hasn't been included in a 3.0.1 release candidate. So I believe if someone cuts a new release candidate from the master branch, we should be able to test it and hopefully confirm that the cicustom issue is no longer blocking the 3.0.1 release.

**Comment by spettinichi** [comment]
The hardcoded numbers in the discs block make this provider extremely difficult to use programmatically. We use a custom module to provision VMs with many server defaults, then in our main.tf we define each VM with only the configuration which varies from default. Dynamically setting disks when the number is hardcoded makes this much harder. We have run into the same issue with the netowrking with ipconfig0, etc. We effectively have to set one NIC in terraform and any additional are created manually after VM creation. I understand this change is required to some extent by changes in Proxmox, but using either a list or perhaps just repeated blocks would make this much more flexible.

Examples:
```
disks {
    scsi [
      {type=passthrough, etc},
      {type=disk, size=60gb, etc}
    ]
  }
  ```
  
  ```
  disks {
    scsi {
      {type=passthrough, etc}
    }
    scsi {
      {type=disk, size=60gb, etc}
    }
  }
  ```

**Reply by hestiahacker** [reply]
I see where you're coming from, but having the number makes it possible to address a specific disk by name instead of having to either specify the index in a list (which may change when the list is modified), or do a lookup based on matching the type + size + other features. Would taking a list of anonymous disks and generating the names in a loop as "scsi+index" (so scsi0, scsi1, etc.) help with this issue? It seems like this might give you the flexibility of dynamically generating things before deployment and have specific indexes that will not change to reference disks after deployment.

**Reply by spettinichi** [reply]
I think implicitly ordering by order could work, but otherwise having a value inside the object which denotes order could also work if we want users to be explicit. 
Implicit ordering in absence of explicit values may be the most flexible while allowing specificity. 
I am willing to spend coding time on this as I am good with Go, but I am not as great with how Terraform works under the hood.

 I have already been working on a change to make NIC allocation more flexible as well, but my first merge request caused an issue and i’m not confident enough in testing to submit yet

**Reply by hestiahacker** [reply]
I'd be happy to help with testing when its ready. I try to spend a few hours on this project every Monday to help test and debug, so testing this would fit right into my schedule. :heart:

**Reply by spettinichi** [reply]
oh excellent! let me take a look over the disk code again tomorrow and see how easily i can adjust that. Do you have discord or anything for easier collaboration?

**Reply by electropolis** [reply]
I think having 

```
 disks {
    scsi {
      scsi0 {
        cdrom {
          //<arguments omitted for brevity...>
        }
      }
      scsi1 {
        disk {
          //<arguments omitted for brevity...>
        }
      }
      scsi2 {
        passthrough {
          //<arguments omitted for brevity...>
        }
      }
      //<arguments omitted for brevity...>
    }
    //<arguments omitted for brevity...>
  }
 ```
 
 is good approach but with small problem. You can't address the `scsi0` as a value. In Yaml annotations it should be something like this
 
 ```
 disks:
   scsi:
     - name: scsi0
        disk:
          size:
          storage:
          ....
```

And then each item of `.disks.scsi` its iterable because it's a list of objects and can be easily looped. 
I don't remember if there is an option to deal with it in HCL like that and will that approach force to do some modifications on variables using `locals`. But right now having current annotations it's impossible to enter `scsi0`.

**Reply by spettinichi** [reply]
@electropolis  so changing it to be like this would work better for what you're trying to do with yaml? Cause it would also help me out for what I am doing

```
 disks {
    scsi {
      {
        name = scsi0
        cdrom {
          //<arguments omitted for brevity...>
        }
      }
      {
        name = scsi1
        disk {
          //<arguments omitted for brevity...>
        }
      }
      {
       name = scsi2
        passthrough {
          //<arguments omitted for brevity...>
        }
      }
      //<arguments omitted for brevity...>
    }
    //<arguments omitted for brevity...>
  }
  ```

**Reply by electropolis** [reply]
Exactly. It would be probably possible
With dynamic argument but I'm not sure. You can also try to put terragrunt on top and organise this in some kind of folder/file structure if you have any idea how terragrunt can be beneficial here.

**Reply by gbrackley** [reply]
@spettinichi I think this might be a balance of getting the normalisation/denormalisation right; and being able to round-trip that with the API.

The nested structure seems to be about:
 - a bus type (scsi/ide/sata/virtio/...)
 - a device type (cdrom/disk/passthrough)
 - a device index/slot 
 - all the arguments that when omitted for brevity belie the residual complexity

Having a structured/specific slot has made the matching of devices for round-tripping with the API easier for the Terraform Provider, but made the resulting terraform less flexible and more structured for the user of the the provider.

I'm still really keen to see an example terraform **module** that allows all disk parameters to be passed through from a terraform script to the terraform **provider**.   Expressing the 'disks' variable in the module, then passing the value provided by the user of that module to the provider seems to be thousands of lines of code.

**Reply by hestiahacker** [reply]
I'm fine with the schema as is and I'm also fine with the schema that @spettinichi [proposed](https://github.com/Telmate/terraform-provider-proxmox/discussions/910#discussioncomment-8801764) and @electropolis [seconded](https://github.com/Telmate/terraform-provider-proxmox/discussions/910#discussioncomment-8806495). I have a slight preference for the latter even through it means updating all my terraform modules that deal with disks... again.

If someone is willing to implement the schema, I say we wait to release 3.0.1. If nobody is volunteering to write that code, then lets move ahead with what we already have in 3.0.1-rc1 so we can get all those bug fixes into an official release.

**Reply by spettinichi** [reply]
we ended up abandoning our module, i've mostly completed rewriting all of our terraform to remove it and be compliant with these changes. I took a look at the code and just don't have the time to learn the ins and outs of the proxmox api enough to re-write it and be confident enough to propose breaking it again. So i now agree with those who want to just cut a release. 

I did find an unrelated bug which seems to already have an open ticket here but it hasn't been addressed https://github.com/Telmate/terraform-provider-proxmox/issues/738 Given this breaking change I had to delete and re-import state, and this bug made that much more tedious than it should have been. I think it'd be nice to get it fixed for this since it is a breaking change, but don't think this should have to hold up the release necessarily.

**Comment by hestiahacker** [comment]
Is there anything we can do to help get an official v3.0.1 release, or is this just a matter of waiting until @mleone87 or @Tinyblargon has time to cut a release?

There are always going to be more tickets to handle, but it'd be nice to have all of these changes go out to people and handle additional feature requests, changes in functionality or bug fixes for a future release (perhaps v3.0.2, which maybe can quickly follow the v3.0.1 release).

The desire to get another release out, even if it's imperfect, was also expressed by @gbrackley [over here](https://github.com/Telmate/terraform-provider-proxmox/issues/884#issuecomment-1899484002)

**Comment by greenship24** [comment]
I have done some testing. I can now reliably deploy a cluster of Fedora CoreOS VMs again. Using proxmox 8.1.3

Not getting any of the usual errors I did on 2.9.x now. Thanks for the hard work.

**Comment by kub3let** [comment]
I just did the upgrade and here are my 0.2$

- qemu-agent vm's with ipv6 only address currently break telmate see https://github.com/Telmate/terraform-provider-proxmox/issues/978
- if using `cicustom` you must specify  `cloudinit_cdrom_storage` now as well, this was not clear and not in the breaking changes, if cicustom is specified but cloudinit_cdrom_storage not it should fail with a error
- with the changes to bools in the disk configuration I think it would make sense to also specify `agent` and maybe other properties as bool instead of 0/1 int as well

**Reply by spettinichi** [reply]
i ran into this creating new VMs with 3.0.1-rc1, which I found is also in an issue here https://github.com/Telmate/terraform-provider-proxmox/issues/727

There needs to be logs that cloudinit_cdrom_storage needs set if any cloudinit stuff is used but it is unset, or it needs to default to the first disk or something. Gotta dig to really figure out because the VM seems to start, it just never gets an IP set or anything. In addition to cicustom, it needs set if setting ipconfig or and login/ssh info.

**Comment by hestiahacker** [comment]
Just checking in again to see if this release can go out. There was a question about changing the schema, but that [is no longer a blocker](https://github.com/Telmate/terraform-provider-proxmox/discussions/910#discussioncomment-9015785) as people have hanged their terraform to match the API in this release.

It'd be good to release the improvements we already have done.

**Reply by kub3let** [reply]
There will be another release candidate https://github.com/Telmate/terraform-provider-proxmox/issues/993

With all the feedback I assume a proper release will take more time.

Nothing prevents you from using the release candidate.
